### PR TITLE
ARROW-10831: [C++][Compute] Implement quantile kernel

### PR DESCRIPTION
**Pull request by cyb70289**
Calculate the exact quantile by storing all values and partition
around quantile point at the end. This may require much memory.

As followup task, an approximate method without storing data points
will be implemented.

### CODE REVIEW

File: cpp/src/arrow/compute/exec.cc [hunk 1]
```diff
@@ -721,19 +720,15 @@ class VectorExecutor : public KernelExecutorImpl<VectorKernel> {
...
-        // XXX: In the case where no outputs are omitted, is returning a 0-length
-        // array always the correct move?
-        return MakeArrayOfNull(output_descr_.type, /*length=*/0).ValueOrDie();
-      }
-    } else {
+    if (kernel_->output_chunked && (HaveChunkedArray(inputs) || outputs.size() > 1)) {
+      return ToChunkedArray(outputs, output_descr_.type);
+    } else if (outputs.size() == 1) {
+      // Outputs have just one element
       return outputs[0];
```
Comment (cyb70289) [inline cpp/src/arrow/compute/exec.cc:727]: `outputs` may be empty vector if input is empty. Use the same `outputs.size()` checking in above code.

**Comment by cyb70289** [comment]
@pitrou , will you review this patch? Thanks.

File: cpp/src/arrow/compute/kernels/aggregate_test.cc [hunk 2]
```diff
@@ -1321,5 +1321,288 @@ TEST_F(TestVarStdKernelIntegerLength, Basics) {
...
+                         O(1.98, 1, 2, 2, 1.5));
+  this->AssertQuantilesAre({"[5]", "[null, null]", "[3, 9, null]", "[8, 1, 2, 0]"},
+                           {0.21, 1}, {O(1.26, 1, 2, 1, 1.5), O(9, 9, 9, 9, 9)});
+
+  this->AssertQuantilesEmpty("[]", {0.5});
+  this->AssertQuantilesEmpty("[null, null, null]", {0.1, 0.2});
+  this->AssertQuantilesEmpty({"[null, null]", "[]", "[null]"}, {0.3, 0.4});
+}
+
+#ifndef __MINGW32__
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/aggregate_test.cc:1480]: Why was this necessary? Can you point to a specific error?

File: cpp/src/arrow/compute/kernels/aggregate_test.cc [hunk 3]
```diff
@@ -1321,5 +1321,288 @@ TEST_F(TestVarStdKernelIntegerLength, Basics) {
...
+
+  this->AssertQuantileIs({"[NaN, -9, 7, Inf]", "[null, NaN]", "[-Inf, NaN, 2, 11]"}, 0.5,
+                         O(4.5, 2, 7, 2, 4.5));
+  this->AssertQuantilesAre({"[null, -9, 7, Inf]", "[NaN, NaN]", "[-Inf, null, 2, 11]"},
+                           {0.3, 0.6}, {O(-3.5, -9, 2, 2, -3.5), O(7, 7, 7, 7, 7)});
+
+  this->AssertQuantilesEmpty("[]", {0.5, 0.6});
+  this->AssertQuantilesEmpty("[null, NaN, null]", {0.1});
+  this->AssertQuantilesEmpty({"[NaN, NaN]", "[]", "[null]"}, {0.3, 0.4});
+}
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/aggregate_test.cc:1506]: What happens if I ask for the median of `[-Inf, Inf]`?

File: cpp/src/arrow/compute/kernels/aggregate_test.cc [hunk 4]
```diff
@@ -1321,5 +1321,288 @@ TEST_F(TestVarStdKernelIntegerLength, Basics) {
...
+  { DOUBLE(a), INTYPE(b), INTYPE(c), INTYPE(d), DOUBLE(e) }
+// output type same as input if only 0 and 1 quantiles are calculated
+#define I(a, b, c, d, e) \
+  { INTYPE(a), INTYPE(b), INTYPE(c), INTYPE(d), INTYPE(e) }
+
+TYPED_TEST_SUITE(TestIntegerQuantileKernel, IntegralArrowTypes);
+TYPED_TEST(TestIntegerQuantileKernel, Basics) {
+  // reference values from numpy
+  // ordered by interpolation method: {linear, lower, higher, nearest, midpoint}
+  this->AssertQuantileIs("[1]", 0.1, O(1, 1, 1, 1, 1));
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/aggregate_test.cc:1446]: Is round-nearest-to-even exercised in the tests below?

**Review by pitrou — COMMENTED** [review]
Thank you @cyb70289 ! Here are some comments.

File: cpp/src/arrow/compute/kernels/aggregate_quantile.cc [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+    const uint64_t lower_index = static_cast<uint64_t>(index);
+    const double fraction = index - lower_index;
+
+    if (lower_index != *last_index) {
+      DCHECK_LT(lower_index, *last_index);
+      std::nth_element(in.begin(), in.begin() + lower_index, in.begin() + *last_index);
+    }
+
+    const double lower_value = static_cast<double>(in[lower_index]);
+    if (fraction == 0 || lower_value == -INFINITY) {
```
[the remark this answers is not in the document]
Comment (cyb70289) [reply cpp/src/arrow/compute/kernels/aggregate_quantile.cc:220]: Numpy returns nan. I didn't follow it. Just returns -inf.

File: cpp/src/arrow/compute/kernels/aggregate_quantile.cc [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+
+// output is at some input data point, not interpolated
+bool IsDataPoint(const QuantileOptions& options) {
+  // some interpolation methods return exact data point
+  if (options.interpolation == QuantileOptions::LOWER ||
+      options.interpolation == QuantileOptions::HIGHER ||
+      options.interpolation == QuantileOptions::NEAREST) {
+    return true;
+  }
+  // return exact data point if quantiles only contain 0 or 1 (follow numpy behaviour)
```
[the remark this answers is not in the document]
Comment (cyb70289) [reply cpp/src/arrow/compute/kernels/aggregate_quantile.cc]: removed

File: cpp/src/arrow/compute/kernels/aggregate_quantile.cc [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+    }
+    for (double q : options.q) {
+      if (q < 0 || q > 1) {
+        ctx->SetStatus(Status::Invalid("Quantile must be between 0 and 1"));
+        return;
+      }
+    }
+
+    // copy all chunks to a buffer, ignore nulls and nans
+    std::vector<CType> in_buffer;
```
[the remark this answers is not in the document]
Comment (cyb70289) [reply cpp/src/arrow/compute/kernels/aggregate_quantile.cc]: done

File: cpp/src/arrow/compute/kernels/aggregate_quantile.cc [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+
+      int64_t index = 0;
+      for (const auto& array : datum.chunks()) {
+        index += CopyArray(in_buffer.data() + index, *array);
+      }
+      DCHECK_EQ(index, in_length);
+
+      // drop nan
+      if (is_floating_type<InType>::value) {
+        const auto& nan_begins = std::partition(in_buffer.begin(), in_buffer.end(),
```
[the remark this answers is not in the document]
Comment (cyb70289) [reply cpp/src/arrow/compute/kernels/aggregate_quantile.cc]: done

File: cpp/src/arrow/compute/kernels/aggregate_quantile.cc [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+      out_type = TypeTraits<InType>::type_singleton();
+    } else {
+      out_type = float64();
+    }
+    auto out_data = ArrayData::Make(out_type, out_length, 0);
+    out_data->buffers.resize(2, nullptr);
+
+    // calculate quantiles
+    if (out_length > 0) {
+      const auto out_bit_width = checked_pointer_cast<NumberType>(out_type)->bit_width();
```
[the remark this answers is not in the document]
Comment (cyb70289) [reply cpp/src/arrow/compute/kernels/aggregate_quantile.cc:108]: Output quantile value type depends on interpolation method, it may be different from CType (input value type).
E.g, linear interpolated quantile from adjacent integers will be double.

File: cpp/src/arrow/compute/kernels/aggregate_quantile.cc [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+      *last_index = lower_index;
+      return lower_value;
+    }
+
+    const uint64_t higher_index = lower_index + 1;
+    DCHECK_LT(higher_index, in.size());
+    if (lower_index != *last_index && higher_index != *last_index) {
+      DCHECK_LT(higher_index, *last_index);
+      // higher value must be the minimal value after lower_index
+      auto min = std::min_element(in.begin() + higher_index, in.begin() + *last_index);
```
[the remark this answers is not in the document]
Comment (cyb70289) [reply cpp/src/arrow/compute/kernels/aggregate_quantile.cc:223]: Next quantile may locate between the same adjacent indices as current quantile. So I moved the correct data point to higher_index at the first time.

File: cpp/src/arrow/compute/kernels/aggregate_test.cc (hunk 3 above)
Comment (cyb70289) [reply cpp/src/arrow/compute/kernels/aggregate_test.cc:1501]: Add a test to exercise this case.
Linear interpolated quantile will be `-Inf`. Numpy returns `NaN`.

File: cpp/src/arrow/compute/kernels/aggregate_test.cc (hunk 4 above)
Comment (cyb70289) [reply cpp/src/arrow/compute/kernels/aggregate_test.cc:1443]: Yes. My old code cutted off 0.5, and I found some test result different from numpy.

File: cpp/src/arrow/compute/kernels/aggregate_test.cc (hunk 2 above)
Comment (cyb70289) [reply cpp/src/arrow/compute/kernels/aggregate_test.cc:1472]: Below are errors from unit test.
First two errors may be caused by index not calculated precisely, pick the wrong side from adjacent data points.
For other failed tests, the printed values are same, looks quantile value is not calculate precisely.

```
D:/a/arrow/arrow/cpp/src/arrow/compute/kernels/aggregate_test.cc:1353: Failure
Expected equality of these values:
  quantiles[j]
    Which is: -9
  numeric_scalar->value
    Which is: -inf
D:/a/arrow/arrow/cpp/src/arrow/compute/kernels/aggregate_test.cc:1353: Failure
Expected equality of these values:
  quantiles[j]
    Which is: inf
  numeric_scalar->value
    Which is: 11
D:/a/arrow/arrow/cpp/src/arrow/compute/kernels/aggregate_test.cc:1361: Failure
Expected equality of these values:
  quantiles[j]
    Which is: -3.5
  numeric_scalar->value
    Which is: -3.5
D:/a/arrow/arrow/cpp/src/arrow/compute/kernels/aggregate_test.cc:1361: Failure
Expected equality of these values:
  quantiles[j]
    Which is: -3.5
  numeric_scalar->value
    Which is: -3.5
D:/a/arrow/arrow/cpp/src/arrow/compute/kernels/aggregate_test.cc:1361: Failure
Expected equality of these values:
  quantiles[j]
    Which is: 7
  numeric_scalar->value
    Which is: 7
D:/a/arrow/arrow/cpp/src/arrow/compute/kernels/aggregate_test.cc:1361: Failure
Expected equality of these values:
  quantiles[j]
    Which is: -3.5
  numeric_scalar->value
    Which is: -3.5
[  FAILED  ] TestFloatingQuantileKernel/0.Floats, where TypeParam = arrow::FloatType (1 ms)
[----------] 1 test from TestFloatingQuantileKernel/0 (1 ms total)

[----------] 1 test from TestFloatingQuantileKernel/1, where TypeParam = arrow::DoubleType
[ RUN      ] TestFloatingQuantileKernel/1.Floats
D:/a/arrow/arrow/cpp/src/arrow/compute/kernels/aggregate_test.cc:1353: Failure
Expected equality of these values:
  quantiles[j]
    Which is: -9
  numeric_scalar->value
    Which is: -inf
D:/a/arrow/arrow/cpp/src/arrow/compute/kernels/aggregate_test.cc:1353: Failure
Expected equality of these values:
  quantiles[j]
    Which is: inf
  numeric_scalar->value
    Which is: 11
D:/a/arrow/arrow/cpp/src/arrow/compute/kernels/aggregate_test.cc:1353: Failure
Expected equality of these values:
  quantiles[j]
    Which is: -3.5
  numeric_scalar->value
    Which is: -3.5
D:/a/arrow/arrow/cpp/src/arrow/compute/kernels/aggregate_test.cc:1353: Failure
Expected equality of these values:
  quantiles[j]
    Which is: -3.5
  numeric_scalar->value
    Which is: -3.5
D:/a/arrow/arrow/cpp/src/arrow/compute/kernels/aggregate_test.cc:1353: Failure
Expected equality of these values:
  quantiles[j]
    Which is: 7
  numeric_scalar->value
    Which is: 7
D:/a/arrow/arrow/cpp/src/arrow/compute/kernels/aggregate_test.cc:1353: Failure
Expected equality of these values:
  quantiles[j]
    Which is: -3.5
  numeric_scalar->value
    Which is: -3.5
[  FAILED  ] TestFloatingQuantileKernel/1.Floats, where TypeParam = arrow::DoubleType (1 ms)
[----------] 1 test from TestFloatingQuantileKernel/1 (1 ms total)

[----------] 1 test from TestInt64QuantileKernel/0, where TypeParam = arrow::Int64Type
[ RUN      ] TestInt64QuantileKernel/0.Int64
[       OK ] TestInt64QuantileKernel/0.Int64 (0 ms)
[----------] 1 test from TestInt64QuantileKernel/0 (0 ms total)

[----------] 2 tests from TestRandomQuantileKernel
[ RUN      ] TestRandomQuantileKernel.Normal
[       OK ] TestRandomQuantileKernel.Normal (7 ms)
[ RUN      ] TestRandomQuantileKernel.Overlapped
D:/a/arrow/arrow/cpp/src/arrow/compute/kernels/aggregate_test.cc:1361: Failure
Expected equality of these values:
  quantiles[j]
    Which is: -47.0185
  numeric_scalar->value
    Which is: -47.0185
[  FAILED  ] TestRandomQuantileKernel.Overlapped (13 ms)
[----------] 2 tests from TestRandomQuantileKernel (20 ms total)

[----------] Global test environment tear-down
[==========] 168 tests from 133 test suites ran. (385 ms total)
[  PASSED  ] 165 tests.
[  FAILED  ] 3 tests, listed below:
[  FAILED  ] TestFloatingQuantileKernel/0.Floats, where TypeParam = arrow::FloatType
[  FAILED  ] TestFloatingQuantileKernel/1.Floats, where TypeParam = arrow::DoubleType
[  FAILED  ] TestRandomQuantileKernel.Overlapped
```

Comment (cyb70289) [reply cpp/src/arrow/compute/kernels/aggregate_test.cc:1472]: Looks mingw32 is still using 80-bit x87 fpu for floating point calculation, while other platforms/compilers uses 64-bit floating point. Might be the reason why many floating point tests failed in arrow unit tests?

Some similar issues:
https://www.raspberrypi.org/forums/viewtopic.php?t=205567
https://stackoverflow.com/questions/13447444/what-difference-between-vs-c-and-mingw-to-implement-double-type

**Comment by cyb70289** [comment]
CI failure is not related.
```
 The following tests FAILED:
	 38 - arrow-s3fs-test (Timeout)
```

File: cpp/src/arrow/compute/kernels/aggregate_test.cc (hunk 3 above)
Comment (pitrou) [reply cpp/src/arrow/compute/kernels/aggregate_test.cc:1501]: Hmm, it's not very important, but `NaN` sounds more logical. What do you think?

File: cpp/src/arrow/compute/kernels/aggregate_test.cc (hunk 2 above)
Comment (pitrou) [reply cpp/src/arrow/compute/kernels/aggregate_test.cc:1472]: Yes, I think that is due to 80-bit FP (double rounding?). However, this doesn't explain the infinity results...

File: cpp/src/arrow/compute/kernels/aggregate_quantile.cc (hunk 9 above)
[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/compute/kernels/aggregate_quantile.cc:108]: Ah, right.

File: cpp/src/arrow/compute/kernels/aggregate_test.cc (hunk 2 above)
Comment (cyb70289) [reply cpp/src/arrow/compute/kernels/aggregate_test.cc:1472]: I didn't investigate deep.
The test array is `[-9, 7, Inf, -Inf, 2, 11]`. So `-inf` and `-9` are two smallest values,  `11` and `inf` are two biggest values.
I guess the error comes from quantile to index conversion, index = quantile * (size - 1). The index is between two data points, -9 and -inf, maybe approaching 0.5, and lead to different result for `nearest interpolation` method.

File: cpp/src/arrow/compute/kernels/aggregate_test.cc (hunk 3 above)
Comment (cyb70289) [reply cpp/src/arrow/compute/kernels/aggregate_test.cc:1501]: Yes, will change.

File: cpp/src/arrow/compute/kernels/aggregate_test.cc (hunk 2 above)
Comment (pitrou) [reply cpp/src/arrow/compute/kernels/aggregate_test.cc:1472]: Ok, that sounds plausible. Let's not worry too much about it. 32-bit x86 platforms are really legacy anyway.

File: cpp/src/arrow/compute/kernels/aggregate_quantile.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (cyb70289) [reply cpp/src/arrow/compute/kernels/aggregate_quantile.cc]: Changed to NaN

File: cpp/src/arrow/compute/kernels/aggregate_test.cc (hunk 3 above)
Comment (cyb70289) [reply cpp/src/arrow/compute/kernels/aggregate_test.cc:1501]: Changed to `NaN`.

File: cpp/src/arrow/compute/kernels/aggregate_test.cc (hunk 4 above)
Comment (cyb70289) [reply cpp/src/arrow/compute/kernels/aggregate_test.cc:1444]: Round-nearest-to-even is covered in tests.

**Comment by cyb70289** [comment]
appveyor ci failure is about s3fs test

**Comment by pitrou** [comment]
I'd rather defer this to 4.0 because `QuantileOptions` will also need to be exposed in Python.

**Comment by cyb70289** [comment]
> I'd rather defer this to 4.0 because `QuantileOptions` will also need to be exposed in Python.

No problem. Thanks.

**Comment by cyb70289** [comment]
Exposed QuantileOptions to Python.
No hurry to merge.
```
In [1]: import pyarrow as pa

In [2]: import pyarrow.compute as pc

In [3]: arr = pa.array([1,2,3,4,5,6,7])

In [4]: option = pc.QuantileOptions(q=(0.1,0.3,0.5), interpolation='nearest')

In [5]: pc.call_function("quantile", [arr], option)
Out[5]: 
<pyarrow.lib.Int64Array object at 0x7f4fc0c345e8>
[
  2,
  3,
  4
]
```

**Review by pitrou — APPROVED** [review]
Thank you very much @cyb70289 !

**Comment by pitrou** [comment]
Rebased and fixed conflict.
